---
title: "RNAseq analysis"
author: ""
date: "`r format(Sys.time(), '%B %d , %Y')`"
output: 
    html_document:
        toc: TRUE
        toc_float: FALSE
        toc_depth: 2
        toc_collapsed: FALSE
        theme: lumen
        number_sections: FALSE
        coding_folding: hide
        code_download: FALSE
editor_options: 
  chunk_output_type: console
params:
  dat.dir: "/fh/fast/_SR/Genomics/user/fwu/project/dmacpherson/20241010_LH00740_0009_A22VGKMLT3/PDX"
---

```{r setup, include=FALSE, dev='CairoPNG'}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning=FALSE, dev="CairoPNG")
knitr::opts_knit$set(root.dir = params$dat.dir)

options(stringsAsFactors = F)
options(scipen = 99)
library(Cairo)
library(ggplot2)
library(dplyr)
library(patchwork)
library(kableExtra)

wd <- params$dat.dir
setwd(wd)

fq.dir <- 'macpherson_lab/20241010_LH00740_0009_A22VGKMLT3/Unaligned/RNAseq_MacPhersonLab_AD_KH_EM_JC_October2024/'
dst.dir <- 'macpherson_lab/20241010_LH00740_0009_A22VGKMLT3/Analysis/PDX'

```

```{r read_args, include=FALSE}
## sample sheet 
ss <- read.csv(paste(wd, 'samples.csv', sep = '/')) %>% 
  mutate(
    fastq_1 = basename(fastq_1),
    fastq_2 = basename(fastq_2)
  ) %>% 
  relocate(fastq_1, fastq_2, .after = last_col())

```

```{r make_tabs, eval=TRUE, include=FALSE}
# Function to create multiple tabs
make_tabs <- function(title, contents){
  cat(title, "\n\n")
  for (cnt in contents){
    print(cnt); 
    cat("\n\n")
    cat('$$\\\\[0.05in]$$')
    cat("\n\n")
    }
  cat("\n\n")
}
```

## Summary

This document describes the analysis of RNAseq data.

$$\\[0.5in]$$

## Sample sheet

```{r sample_sheet, eval=TRUE, include=TRUE}

ss %>%  
  dplyr::rename(Sample = sample_id) %>% 
  DT::datatable(
  class = 'cell-border stripe', 
  rownames = T,  
  filter = 'top', 
  editable = F, 
  extensions = 'Buttons', 
  options = list(
    pageLength = 6, 
    dom = 'Bfrtip',
    buttons = c('csv')
  ))
```

$$\\[0.5in]$$

## Differential gene expression

### DE Summary

feature = gene; FDR = BH-corrected p-value; FC = fold change\
Note:\
Statistics of DE genes defined by an alternative threshold are also listed between columns *features.up2* and *FC.cutoff2*.

```{r}

read.delim('Analysis/differential_expression/DE_summary.txt') %>% 
  relocate(control.samples, test.samples, .after = last_col()) %>% 
  DT::datatable(
  class = 'cell-border stripe', 
  rownames = T,  
  filter = 'top', 
  editable = F, 
  extensions = 'Buttons', 
  options = list(
    pageLength = 6, 
    dom = 'Bfrtip',
    buttons = c('csv')
  ))

```

$$\\[0.2in]$$

### DE plots {.tabset .tabset-pills}

DE genes were defined by the first threshold in section *DE Summary* (between columns *features.up* and *FC.cutoff*) .\
Note:\
PCA is based on top 500 variable genes (NOT on chr X, chr Y or chr M).\
See PCA plot of all samples in *differential_expression/all_samples.PCA.pdf*\
The PCA plot of samples relevant to a specific comparison are also shown below.\
\

```{r v1, eval=TRUE, include=TRUE, results='asis', fig.cap=c('Fig. Differential expression analysis. (A) PCA plot. (B) MA/MD plot. (C) Volcano plot'), fig.width=8, fig.height=4}
dge <- readRDS('Analysis/differential_expression/de.rds')

for (i in 1:length(dge)){
  plist <- dge[[i]]$plots
  p <- wrap_plots(plist, nrow = 1) + plot_annotation(tag_levels = 'A')
  
  make_tabs(
    paste("####", names(dge)[i]),
    list(p)
  )
}
```

$$\\[0.5in]$$

## Delivered files

### Fastq files

`r fq.dir`

$$\\[0.2in]$$

### Analysis results

`r dst.dir` includes the following folders/files.

#### [differential_expression/](./differential_expression/)

*all_samples.PCA.pdf* - PCA plot of all samples.\
*all_samples.raw_counts.txt* - Raw gene-level counts of all protein-coding and lncRNA genes in all samples. Genes with zero counts across all samples are excluded.\

Each comparison has a subfolder (see section *DE summary* for more details). The text file has edgeR output (logFC, logCPM, p-value, BH-adjusted FDR), normalized expression in RPKM and CPM. In column is.sig of the text file, each gene is assigned 1 for up-regulation, -1 for down-regulation and 0 for no significant difference as defined by the cutoffs of FDR \< 0.05 and fold change \> 1.5 at either direction. PCA, MD/MA and volcano plots are in the same subfolder and also included in this report under section *DE plots*.\

$$\\[0.1in]$$

#### [featureCounts/](.featureCounts/)

featureCounts generated gene-level counts that were latter used for differential gene expression.

$$\\[0.1in]$$

#### [MultiQC/](./MultiQC)

multiqc_report.html -- MultiQC html report aggregating results from individual QC tools and STAR aligner.

$$\\[0.1in]$$

#### [QC/](./QC)

Results from multiple QC tools.\

##### [FastQC/](./FastQC)

QC by [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/).\
\

##### [RNA-SeQC/](./RNA-SeQC)

QC by [RNA-SeQC](https://doi.org/10.1093/bioinformatics/btab135).\
\

##### [RSeQC/](./RSeQC)

QC by [RSeQC](http://doi.org/10.1093/bioinformatics/bts356)

$$\\[0.1in]$$

#### [STAR/](./STAR)

Output from STAR aligner and XenofilteR.

**hg38/** -- STAR alignment to hg38. Genomic alignment bam files and bam index files are named as *[sample name].bam* and *[sample name].bam.bai*.\
**hg38_filtered/** -- XenofilteR processed bam files where reads of mouse origin are removed. Files are names as *[sample name]\_filtered.bam* and *[sample name]\_filtered.bam.bai.\
***mm10/** *-- STAR alignment to mm10. Genomic alignment bam files and bam index files are named as [sample name].bam* and *[sample name].bam.bai*.\

$$\\[0.1in]$$

##  {.unnumbered .unlisted}

$$\\[0.5in]$$

## Methods

[STAR v2.7.7a](https://pubmed.ncbi.nlm.nih.gov/23104886/) with 2-pass mapping was used to align paired-end reads to human hg38 assembly and GENCODE annotation v38. [FastQC 0.11.9](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/), [RNA-SeQC 2.3.4](https://pubmed.ncbi.nlm.nih.gov/22539670/), [RSeQC 4.0.0](https://pubmed.ncbi.nlm.nih.gov/22743226/) were used to check various QC metrics including insert fragment size, read quality, read duplication rates, rRNA rates, gene body coverage and read distribution in different genomic regions. [XenofilteR](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2353-5) was used to remove reads of mouse origin following authors-recommended parameters. [featureCounts](https://academic.oup.com/bioinformatics/article/30/7/923/232889) in Subread v2.0.0 was used to quantify gene-level expression in unstranded fashion. Bioconductor package [edgeR 3.36.0](https://academic.oup.com/bioinformatics/article/26/1/139/182458) was used to detect differential gene expression between sample groups. Genes with low expression were excluded using edgeR function filterByExpr with min.count = 10 and min.total.count = 15. The filtered expression matrix was normalized by [TMM method](https://doi.org/10.1186/gb-2010-11-3-r25) and subject to significance testing using quasi-likelihood pipeline implemented in edgeR. A gene was deemed differentially expressed if log2 fold change was above 1.5 in either direction and Benjamini-Hochberg adjusted p-values was less than 0.05.

##  {.unnumbered .unlisted}

$$\\[0.5in]$$

## Acknowledgement

(For publication, please include the following acknowledgement.) This research was supported by the Genomics & Bioinformatics Shared Resource of the Fred Hutch/University of Washington/Seattle Children's Cancer Consortium ([P30 CA015704](https://www.cancerconsortium.org/about.html)).

$$\\[0.5in]$$

##  {.unnumbered .unlisted}

```{r exit_knit}
knitr::knit_exit()
```

# Analyses {.unnumbered .unlisted}

## Create a sample sheet and comparison table

```{r}
meta <- xlsx::read.xlsx('../_user/241011_BulkRNAseq_analysisrequest.xlsx', sheetIndex = 1) %>% 
  dplyr::select(Sample_name:Analysis.group.name) %>% 
  filter(!is.na(Sample_name))
fq <- read.csv('fq.csv')


## sample sheet ####
ss <- fq %>% 
  left_join(
    meta, by = c('sample_id' = 'Sample_name')
  ) %>% 
  mutate(
    sample_group = Analysis.group,
    sample_group_name = Analysis.group.name,
    analysis_type = case_when(
      Species == 'mouse' ~ 'mm10',
      Species == 'human' & Cell.line.or.tumor == 'tumor' ~ 'PDX',
      Species == 'human' ~ 'hg38',
      TRUE ~ 'TBD'
    )
  ) %>% 
  dplyr::select(
    sample_id, fastq_1, fastq_2, sample_group, sample_group_name, analysis_type
  )
ss %>% write.table('samples.csv', sep = ',', quote = F, row.names = F)



## comparison table ####
cmp <- tribble(
  ~control.group, ~test.group,
  'F1', 'F2',
  'F3', 'F4',
  'F5', 'F6',
  'F1', 'F3',
  'F1', 'F5'
) %>% 
  rowwise() %>% 
  mutate(
    cg = paste(control.group, collapse = '+'),
    tg = paste(test.group, collapse = '+'),
    out.prefix = paste0(tg, '_vs_', cg, '/', tg, '_vs_', cg),
    plot.title = paste(tg, 'vs', cg, '(reference)')
    ) %>% 
  dplyr::select(-c(cg,tg)) # comparisons of interest

cmp %>% saveRDS('comparisons.rds')

## add more comparisions
cmp2 <- tribble(
  ~control.group, ~test.group,
  'F1', c('F3','F5'), 
  'F2', 'F4',
  'F2', 'F6',
  'F2', c('F4', 'F6')
) %>% 
  rowwise() %>% 
  mutate(
    cg = paste(control.group, collapse = '+'),
    tg = paste(test.group, collapse = '+'),
    out.prefix = paste0(tg, '_vs_', cg, '/', tg, '_vs_', cg),
    plot.title = paste(tg, 'vs', cg, '(reference)')
    ) %>% 
  dplyr::select(-c(cg,tg)) # comparisons of interest

cmp2 %>% saveRDS('comparisons2.rds')
```

## Run nextflow pipeline

```{bash}
shdir=~/github/rnaseq/local/bin
bash $shdir/fqdir_to_csv.sh path/to/fastq/
# (optional) adding additional metadata to sample sheet
cp fq.csv samples.csv

# edit params.json and/or nextflow.config as neccessary
bash run_rnaseq.sh -resume

## DE rerun
_rnaseq/bin/local/sbatch_R.sh differential_expression.r input=samples.csv comparison=comparisons2.rds count.dir=Analysis/featureCounts gene.txt=/fh/fast/_SR/Genomics/proj/fwu/reference/hg38/genes/gencode.v38.annotation.proteinCoding_lncRNA.genes.txt length.col=full_exome strand=0 fdr=0.05 fc=1.5 fdr2=0.01 fc2=2

```

# Deliver data

```{bash}
user=dmacpherson
fc=20241010_LH00740_0009_A22VGKMLT3
cd /fh/fast/_SR/Genomics/user/fwu/project/$user/$fc
log=/fh/fast/_SR/Genomics/user/fwu/project/log/$user.$fc.rsync.log


## 24.11.15
dst=/shared/ngs/illumina/macpherson_lab/$fc/
ls -ldh $dst
dstgroup=$(ls -ldh $dst | cut -f 4 -d ' ')
echo $dstgroup

dst=$dst/Analysis
[[ -d $dst ]] || mkdir -p $dst

## PDX
src=/fh/fast/_SR/Genomics/user/fwu/project/dmacpherson/20241010_LH00740_0009_A22VGKMLT3/Analysis/
subdir=PDX
rsync -va --chown=solexa:$dstgroup --chmod=775 $subdir/241010_RNAseq.html $dst/$subdir/
rsync -vAXEWHhrLog --chown=solexa:$dstgroup --chmod=775 --no-compress --progress --exclude-from=to_exclude.txt $src/$subdir/differential_expression_new/ $dst/$subdir/differential_expression/ | tee -a $log


## Appendix
sudo pmanage restore --path $dst --group $dstgroup
sudo pmanage read_other --path $dst
ypmatch -k $user auto.illumina
```

# Appendix

```{r}
library(plotly)

ggplotly(p, tooltip = "text")

subplot(style(figs$seq_depth$AR, showlegend=F), figs$seq_depth$H3K27me3, margin = 0.1) # suppress legends

```
