---
title: "RNAseq analysis"
author: ""
date: "`r format(Sys.time(), '%B %d , %Y')`"
output: 
    html_document:
        toc: TRUE
        toc_float: FALSE
        toc_depth: 2
        toc_collapsed: FALSE
        theme: lumen
        number_sections: FALSE
        coding_folding: hide
        code_download: FALSE
editor_options: 
  chunk_output_type: console
params:
  dat.dir: "./"
---

```{r setup, include=FALSE, dev='CairoPNG'}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning=FALSE, dev="CairoPNG")
knitr::opts_knit$set(root.dir = params$dat.dir)

options(stringsAsFactors = F)
options(scipen = 99)
library(Cairo)
library(ggplot2)
library(dplyr)
library(patchwork)
library(kableExtra)

wd <- params$dat.dir
setwd(wd)

```



## Summary

This document describes the analysis of RNAseq data.

$$\\[0.5in]$$

## {.unnumbered .unlisted}

```{r make_tabs, eval=TRUE, include=FALSE}
# Function to create multiple tabs
make_tabs <- function(title, contents){
  cat(title, "\n\n")
  for (cnt in contents){
    print(cnt); 
    cat("\n\n")
    cat('$$\\\\[0.05in]$$')
    cat("\n\n")
    }
  cat("\n\n")
}
```

```{r plot_functions}
## PCA
plot_pca <- function(y, var.genes = NULL, color = NULL, plot.title = '', sample.label = T, feature.length = 'gene_length'){
    options(stringsAsFactors = F)
    require(ggrepel)
    require(edgeR)
    

    if('chrom' %in% colnames(y$genes)){
        y <- y[!y$genes$chrom %in% c('chrX', 'chrY', 'chrM'), ]
    }
    
    log2rpkm <- rpkm(y, gene.length = feature.length, normalized.lib.sizes = T, log = T)
    
    if (!is.null(var.genes)){
        keep <- rank(-apply(log2rpkm, 1, var)) < var.genes
        log2rpkm <- log2rpkm[keep,]
    }
    
    ## Run PCA
    pca <- prcomp(t(log2rpkm), center = T, scale = T)
    
    ## Scree plot
    pca.variance.prop <- (pca$sdev^2)/sum(pca$sdev^2)*100
    
    ## PC1 vs PC2
    df <- cbind(pca$x[,1:2],data.frame(label=rownames(pca$x)))
    if(is.null(color)){
        df$color <- y$samples$group
    }else{
        df$color <- color
    }
    
    if(length(unique(df$color))>1){
        p <- ggplot(df,aes(x=PC1,y=PC2,label=label,color=color))
    }else{
        p <- ggplot(df,aes(x=PC1,y=PC2,label=label))
    }
    p <- p +
        geom_point(shape = 1)
    
    if(sample.label){
        p <- p +
            geom_text_repel(size = 2.4, color = 'black', position = 'jitter',max.overlaps = 80)
    }
    
    p <- p +
        labs(
            title = plot.title,
            color = '', 
            x = paste0('PC1 (',round(pca.variance.prop[1],1),'%)'),
            y = paste0('PC2 (',round(pca.variance.prop[2],1),'%)')
        )+
        theme_bw()+
        theme(
            legend.position = 'top'
        )

    ## return data
    return(p)
    
}

## Plot MD
plot_MD <- function(df, plot.title = ""){
    require(ggplot2)
    
    p <- ggplot(df, aes(x = logCPM, y = logFC, color=factor(is.sig)))+
        geom_hline(yintercept = 0)+
        geom_point(size = 0.2)+
        scale_color_manual(
            '',
            values = c("-1" = "blue", "0" = "gray", "1" = "red"),
            breaks = c(-1,0,1), labels = c('Down', 'No Sig.', 'Up')
        )+
        labs( 
            x = "Average log CPM", 
            y = "log-fold-of-change",
            title = plot.title
        )+
        theme_bw()+
        theme(
            text = element_text(size = 8),
            legend.position = 'top'
        )
    
    return(p)
} 

## Plot volcano 
plot_volcano <- function(df, plot.title = ""){
    require(ggplot2)
    
    p <- ggplot(df,aes(x=logFC,y=-log10(FDR),color=factor(is.sig)))+
        geom_point(size = 0.2)+ 
        scale_color_manual(
            values = c('-1'='blue','0'='gray','1'='red'),
            breaks = c('-1','0','1'),
            labels = c('Down','No Sig.','Up'),
            drop = T
        )+
        labs(
            x='log2(fold change)',
            y='-log10(FDR)',
            color='',
            title = plot.title
        )+
        theme_bw()+
        theme(
            text = element_text(size = 8),
            legend.position = 'top'
        )
    
    return(p)
}

## recompute is.sig2
recal_sig <- function(txt, col.sig, fdr, fc){
    de <- read.delim(txt)
    de[,col.sig] <- sign(de$logFC) * (abs(de$logFC) > log2(fc)) * (de$FDR < fdr)
    write.table(de, txt, sep = '\t', quote = F, row.names = F)
    return(
        data.frame(
            source.file = basename(txt),
            modify.col = col.sig,
            features.test = nrow(de),
            features.up = sum(de[,col.sig] %in% 1),
            features.down = sum(de[,col.sig] %in% -1),
            FDR.cutoff = fdr,
            FC.cutoff = fc
        )
    )
}

```

```{r read_input}
## sample sheet 
ss <- read.csv(paste(wd, 'samples.csv', sep = '/')) %>% 
  mutate(
    fastq_1 = basename(fastq_1),
    fastq_2 = basename(fastq_2)
  ) %>% 
  relocate(fastq_1, fastq_2, .after = last_col())

```

```{r run_de}
de.rds <- 'de.rds'
if (file.exists(de.rds)){
  de.list <- readRDS(de.rds) # by conp, then by cmp
}

if (exists('de.list')){
  de_sum <- bind_rows(lapply(de.list, function(de){de$summary})) %>% 
    dplyr::relocate(output.folder, control.group, test.group, features.tested:FC.cutoff, control.samples:test.samples)
  de_sum %>% 
    write.table('DE_summary.txt', sep = '\t', quote = F, row.names = F)
  
 
  de_plots <- lapply(
    de.list, function(de){
      mapply(function(ide, prefix){
        y <- ide$y
        df <- ide$df
        out.prefix <- basename(prefix)
        plot.title <- gsub('_vs_', ' vs ', out.prefix)
        list(
          PCA = plot_pca(
            y, 
            out.prefix, 
            color = y$samples$sample_group, 
            sample.label = T, 
            plot.title = "", 
            var.genes = 500, 
            feature.length = "length"),
          MD = plot_MD(
            df, out.prefix = out.prefix, 
            plot.title = plot.title
          ),
          volcano = plot_volcano(
            df, out.prefix = out.prefix, 
            plot.title = plot.title
          )
        )
        
      }, de, names(de), SIMPLIFY = F)
    }
  )
  
}
```

## Sample sheet
```{r sample_sheet}

ss %>%  
  DT::datatable(
  class = 'cell-border stripe', 
  rownames = T,  
  filter = 'top', 
  editable = F, 
  extensions = 'Buttons', 
  options = list(
    pageLength = 6, 
    dom = 'Bfrtip',
    buttons = c('csv')
  ))
```

$$\\[0.5in]$$


## Differential gene expression

### DE Summary
feature = gene; FDR = BH-corrected p-value; FC = fold change  \
Note: \
Statistics of DE genes are summarized based on two thresholds -- the first set between columns *features.up* and *FC.cutoff* and the second set between columns *features.up2* and *FC.cutoff2*.

```{r}

read.delim('Analysis/differential_expression/DE_summary.txt') %>% 
  relocate(control.samples, test.samples, .after = last_col()) %>% 
  DT::datatable(
  class = 'cell-border stripe', 
  rownames = T,  
  filter = 'top', 
  editable = F, 
  extensions = 'Buttons', 
  options = list(
    pageLength = 6, 
    dom = 'Bfrtip',
    buttons = c('csv')
  ))

```

$$\\[0.2in]$$

### DE plots {.tabset .tabset-pills}

DE genes displayed in the plots below and in the delivered pdf files  were defined by the first threshold in section *DE Summary* (between columns *features.up* and *FC.cutoff*) .  \
Note: \
PCA is based on top 500 variable genes (NOT on chr X, chr Y or chr M). \
See PCA plot of all samples in *differential_expression/all_samples.PCA.pdf* \
The PCA plot of samples relevant to a specific comparison are also shown below. \
\

```{r v1, eval=TRUE, include=TRUE, results='asis', fig.cap=c('Fig. Differential expression analysis. (A) PCA plot. (B) MA/MD plot. (C) Volcano plot'), fig.width=8, fig.height=4}
dge <- readRDS('Analysis/differential_expression/de.rds')

for (i in 1:length(dge)){
  plist <- dge[[i]]$plots
  p <- wrap_plots(plist, nrow = 1) + plot_annotation(tag_levels = 'A')
  
  make_tabs(
    paste("####", names(dge)[i]),
    list(p)
  )
}
```


$$\\[0.5in]$$

## Delivered files

### Fastq files

`r fq.dir`

$$\\[0.2in]$$

### Analysis results

`r dst.dir` includes the following folders/files.\
\

#### [differential_expression/](./differential_expression/)

*all_samples.PCA.pdf* - PCA plot of all samples.\
*all_samples.raw_counts.txt* - Raw gene-level counts of all protein-coding and lncRNA genes in all samples. Genes with zero counts across all samples are excluded. \

Each comparison has a subfolder (see section *DE summary* for more details). The text file has edgeR output (logFC, logCPM, p-value, BH-adjusted FDR), normalized expression in RPKM and CPM. In column is.sig of the text file, each gene is assigned 1 for up-regulation, -1 for down-regulation and 0 for no significant difference as defined by the cutoffs of FDR \< 0.05 and fold change \> 1.5 at either direction. PCA, MD/MA and volcano plots are in the same subfolder and also included in this report under section *DE plots*.\

$$\\[0.1in]$$

#### [featureCounts/](.featureCounts/)

featureCounts generated gene-level counts that were latter used for differential gene expression.

$$\\[0.1in]$$

#### [MultiQC/](./MultiQC)

multiqc_report.html -- MultiQC html report aggregating results from individual QC tools and STAR aligner.

$$\\[0.1in]$$

#### [QC/](./QC)
Results from multiple QC tools.\

##### [fastqc/](./FastQC)

QC by [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/).\
\

##### [rnaseqc/](./RNA-SeQC)

QC by [RNA-SeQC](https://doi.org/10.1093/bioinformatics/btab135).\
\

##### [rseqc/](./RSeQC)

QC by [RSeQC](http://doi.org/10.1093/bioinformatics/bts356)

$$\\[0.1in]$$


#### [STAR/](./STAR)

Output from STAR aligner.
Genomic alignment bam files and bam index files are named as *[sample name].bam* and *[sample name].bam.bai*.

$$\\[0.1in]$$


##  {.unnumbered .unlisted}

$$\\[0.5in]$$

## Methods

[STAR v2.7.7a](https://pubmed.ncbi.nlm.nih.gov/23104886/) with 2-pass mapping was used to align paired-end reads to human hg38 assembly and GENCODE annotation v38. [FastQC 0.11.9](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/), [RNA-SeQC 2.3.4](https://pubmed.ncbi.nlm.nih.gov/22539670/), [RSeQC 4.0.0](https://pubmed.ncbi.nlm.nih.gov/22743226/) were used to check various QC metrics including insert fragment size, read quality, read duplication rates, rRNA rates, gene body coverage and read distribution in different genomic regions. [featureCounts](https://academic.oup.com/bioinformatics/article/30/7/923/232889) in Subread v2.0.0 was used to quantify gene-level expression in unstranded fashion. Bioconductor package [edgeR 3.36.0](https://academic.oup.com/bioinformatics/article/26/1/139/182458) was used to detect differential gene expression between sample groups. Genes with low expression were excluded using edgeR function filterByExpr with min.count = 10 and min.total.count = 15. The filtered expression matrix was normalized by [TMM method](https://doi.org/10.1186/gb-2010-11-3-r25) and subject to significance testing using quasi-likelihood pipeline implemented in edgeR. A gene was deemed differentially expressed if log2 fold change was above 1.5 in either direction and Benjamini-Hochberg adjusted p-values was less than 0.05.

$$\\[0.5in]$$

## Acknowledgement
(For publication, please include the following acknowledgement.)
This research was supported by the Genomics & Bioinformatics Shared Resource of the Fred Hutch/University of Washington/Seattle Children's Cancer Consortium ([P30 CA015704](https://www.cancerconsortium.org/about.html)).

$$\\[0.5in]$$

